### DISCUSSION TITLE
Does quartz can migrate database automatically?

### DISCUSSION THREAD

**Discussion by sergeyshaykhullin** [quartznet/quartznet#925] [General] [answered]
Should i execute migration scripts manually?
How can i use quartz with ci/cd or development environment?

**Comment by lahma** [comment]
Generic case for new environment is that you just run the schema scripts found [here](https://github.com/quartznet/quartznet/tree/master/database/tables) to build the tables Quartz needs.

If you have the need to upgrade older version to newer, you should follow the schema upgrade scripts found [here](https://github.com/quartznet/quartznet/tree/master/database). For example upgrading from 2.0 to 3.1 would mean that you need to run the first three scripts from there.

**Comment by sergeyshaykhullin** [comment]
@lahma  I have an idea:

How about to use ef core migration builder to setup migration scripts:
1. Add posibility to migrate in-process or out of process with global tool/`dotnet ef script`
2. Plug and play nuget distribution
3. Ef core tracks and generates schema changes for us without `schema_26_to_30.sql` etc.
4. In-code table prefix/schema changes
5. Schema versioning out the box
6. Ef Core supports all quartz officially supported databases
7. Data migration out the box with Ef Core MigrationOperation

### Vision:
1. Nuget packages per provider:
- Quartz.EntityFrameworkCore.Migrations.PostgreSQL (with Quartz version e.g. 3.1.0)
- ...

2. Package contains extension on ef core model builder: `.AddQuartzPostgreSQLOrSomethingLikeThat("specify prefix", "schema", ...)`

3. `dotnet ef migrations add AddQuartz ...`

4. `dotnet ef database update`/`databaseContext.Database.Migrate()`/`dotnet ef script`/global tool to migrate database before/after deploy new version

### Results
- Quartz core still using ADO.NET providers with no EF dependencies
- EF (if used, but a lot of projects is using it) responsible only for migrations with one line setup
- Or you can use old way and execute sql scripts manually

### Example code
```cs
// job details entity
modelBuilder.Entity("job_details", builder =>
{
	builder.ToTable("job_details", "quartz");

	builder.HasKey("sched_name", "job_name", "job_group");

	builder.Property("sched_name")
		.HasColumnType("text")
		.IsRequired();

	builder.Property("job_name")
		.HasColumnType("text")
		.IsRequired();

	builder.Property("job_group")
		.HasColumnType("text")
		.IsRequired();

	builder.Property("description")
		.HasColumnType("text")
		.IsRequired();

	builder.Property("job_class_name")
		.HasColumnType("text")
		.IsRequired();

	builder.Property("is_durable")
		.HasColumnType("bool")
		.IsRequired();

	builder.Property("is_nonconcurrent")
		.HasColumnType("bool")
		.IsRequired();

	builder.Property("is_update_data")
		.HasColumnType("bool")
		.IsRequired();

	builder.Property("requests_recovery")
		.HasColumnType("bool")
		.IsRequired();

	builder.Property("job_data")
		.HasColumnType("bytea");

	builder.HasIndex("requests_recovery")
		.HasDatabaseName("idx_qrtz_j_req_recovery");
});

```

**Comment by sergeyshaykhullin** [comment]
@lahma I created PostgreSQL proof-of-concept migrations for Quartz. It is boring, but possible https://github.com/appany/AppAny.Quartz.EntityFrameworkCore.Migrations.PostgreSQL

Also you can generate scripts in database folder automagically from ef core schema. So EFCore model will be source of truth

**Comment by lahma** [comment, marked as answer]
Hey thanks for bringing the idea to the table! 👍 

Currently I don't think that I can/will take the maintenance burden of per-database NuGet packages (whether it's schema or driver delegate). So far Quartz has been quite slim to use with different databases, no special integration packages needed - just the driver you want to use. Story might be a bit different if Quartz would use NHibernate/EF or similar. I'm unsure how well EF migrations would handle a scenario where for example we want to change the primary keys etc, might be most efficient to use per-database optimized migration script.

I think community could benefit of such helpers if you find the use case important for, it's probably for some others too.

**Reply by sergeyshaykhullin** [reply]
I will maintain this package for our needs

Thanks
